disnake/shard.py

```diff
@@ -329,29 +331,58 @@ class AutoShardedClient(Client):
     """
 
     if TYPE_CHECKING:
+        shard_ids: Optional[List[int]]
         _connection: AutoShardedConnectionState
-
-    def __init__(
-        self, *args: Any, loop: Optional[asyncio.AbstractEventLoop] = None, **kwargs: Any
-    ) -> None:
-        kwargs.pop("shard_id", None)
-        self.shard_ids: Optional[List[int]] = kwargs.pop("shard_ids", None)
-        super().__init__(*args, loop=loop, **kwargs)
-
-        if self.shard_ids is not None:
-            if self.shard_count is None:
-                raise ClientException(
-                    "When passing manual shard_ids, you must provide a shard_count."
-                )
-            elif not isinstance(self.shard_ids, (list, tuple)):
-                raise ClientException("shard_ids parameter must be a list or a tuple.")
-
-        # instead of a single websocket, we have multiple
-        # the key is the shard_id
-        self.__shards = {}
-        self._connection._get_websocket = self._get_websocket
-        self._connection._get_client = lambda: self
-        self.__queue = asyncio.PriorityQueue()
+        __shards: Dict[int, Shard]
+        __queue: asyncio.PriorityQueue
+
+        def __init__(
+            self,
+            *,
+            asyncio_debug: bool = False,
+            loop: Optional[asyncio.AbstractEventLoop] = None,
+            shard_ids: List[int] = None,  # instead of Client's shard_id: Optional[int]
+            shard_count: Optional[int] = None,
+            enable_debug_events: bool = False,
+            connector: Optional[aiohttp.BaseConnector] = None,
+            proxy: Optional[str] = None,
+            proxy_auth: Optional[aiohttp.BasicAuth] = None,
+            assume_unsync_clock: bool = True,
+            max_messages: Optional[int] = 1000,
+            application_id: Optional[int] = None,
+            heartbeat_timeout: float = 60.0,
+            guild_ready_timeout: float = 2.0,
+            allowed_mentions: Optional[AllowedMentions] = None,
+            activity: Optional[BaseActivity] = None,
+            status: Optional[Union[Status, str]] = None,
+            intents: Intents = ...,
+            chunk_guilds_at_startup: Optional[bool] = None,
+            member_cache_flags: MemberCacheFlags = None,
+            cache_application_command_permissions: bool = True,
+        ):
+            ...
+
+    else:
```

Comment:
This makes tools skip type-checking the `__init__` implementation, which isn't ideal. Overloads could be a good alternative (strict type-checkers don't like overlapping overloads, but there are several already so might as well add another one)

Reply:
Replaced with overlapping overload jank in c676630624e26412dc65245ee7c41c6c82abcf17.